File: src/main/scala/com/github/skozlov/async/Deadline.scala
package com.github.skozlov.async

import java.time.{Clock, Instant}
import java.time.temporal.ChronoUnit.NANOS
import scala.concurrent.duration.Duration

sealed trait Deadline extends Ordered[Deadline] {
    def toTimeout: Duration

    def isOver: Boolean = toTimeout <= Duration.Zero
}

object Deadline {
    case object Inf extends Deadline {
        override def compare(that: Deadline): Int = {
            if (this == that) 0 else 1
        }

        override val toTimeout: Duration = Duration.Inf
    }

    case object MinusInf extends Deadline {
        override def compare(that: Deadline): Int = {
            if (this == that) 0 else -1
        }

        override val toTimeout: Duration = Duration.MinusInf
    }

    case class At(instant: Instant, clock: Clock) extends Deadline {
        override def compare(that: Deadline): Int = that match {
            case Inf => -1
            case MinusInf => 1
            case At(thatInstant, _) => this.instant compareTo thatInstant
        }

        override def toTimeout: Duration = Duration.fromNanos(clock.instant().until(instant, NANOS))
    }

    implicit def fromTimeout(timeout: Duration)(implicit clock: Clock): Deadline = timeout match {
        case Duration.Inf => Inf
        case Duration.MinusInf => MinusInf
        case _ => At(clock.instant() plusNanos timeout.toNanos, clock)
    }
}

File: src/main/java/com/github/skozlov/async/deadline/Deadline.java
package com.github.skozlov.async.deadline;

import lombok.NonNull;
import lombok.RequiredArgsConstructor;

import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import java.util.Objects;

import static java.lang.Math.max;
import static java.time.temporal.ChronoUnit.NANOS;

@RequiredArgsConstructor
public class Deadline implements Comparable<Deadline> {
  @NonNull
  private final Instant instant;

  @NonNull
  private final Clock clock;

  public @NonNull Instant toInstant() {
    return instant;
  }

  public boolean isPassed() {
    return !clock.instant().isBefore(instant);
  }

  public @NonNull Duration getTimeLeft() {
    return Duration.ofNanos(max(0L, clock.instant().until(instant, NANOS)));
  }

  @Override
  public int compareTo(@NonNull Deadline that) {
    return this.instant.compareTo(that.instant);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Deadline deadline = (Deadline) o;
    return Objects.equals(instant, deadline.instant) && Objects.equals(clock, deadline.clock);
  }

  @Override
  public int hashCode() {
    return Objects.hash(instant, clock);
  }

  @Override
  public String toString() {
    return "Deadline(" + instant + ")";
  }
}
